pkg/data_cache/src/worker/indexable_mem_table.rs

```diff
@@ -0,0 +1,385 @@
+}
+
+impl IndexableMemTable {
+    pub fn try_new(
+        schema: SchemaRef,
+        partitions: Vec<Vec<RecordBatch>>,
+        indices: Vec<u64>,
+    ) -> datafusion::common::Result<Self> {
+        for batches in partitions.iter().flatten() {
+            let batches_schema = batches.schema();
+            if !schema.contains(&batches_schema) {
+                error!(
+                    "mem table schema does not contain batches schema. \
+                        Target_schema: {schema:?}. Batches Schema: {batches_schema:?}"
+                );
+                return plan_err!("Mismatch between schema and batches");
+            }
+        }
+
+        Ok(Self {
+            schema,
+            batches: partitions.into_iter().flatten().collect(),
+            indices,
+        })
+    }
+
+    pub async fn load(
+        t: Arc<dyn TableProvider>,
+        _output_partitions: Option<usize>,
+        _state: &SessionState,
+        start_index: u64,
+    ) -> datafusion::common::Result<Self> {
+        let schema = t.schema();
+        let exec = t.scan(_state, None, &[], None).await?;
+
+        let mut data: Vec<RecordBatch> = vec![];
+        let mut indices: Vec<u64> = vec![];
+
+        let mut start_index = start_index;
+        let stream = exec.execute(0, _state.task_ctx())?;
+        let _ = stream
+            .map_ok(|batch| {
+                indices.push(start_index);
+                start_index += batch.num_rows() as u64;
+                data.push(batch);
+            })
+            .collect::<Vec<_>>()
+            .await;
+
+        info!("Number of batches: {}", data.len());
+
+        // let mut exec = MemoryExec::try_new(&[data], Arc::clone(&schema), None)?;
+        // if let Some(cons) = constraints {
+        //     exec = exec.with_constraints(cons.clone());
+        // }
+        //
+        // if let Some(num_partitions) = output_partitions {
+        //     // TODO: handle repartioning
+        // }
+        IndexableMemTable::try_new(Arc::clone(&schema), vec![data], indices)
+    }
+}
+
+async fn fetch_partitions(
+    batches: Vec<RecordBatch>,
+    indices: &[u64],
+    start: u64,
+    end: u64,
+) -> Vec<RecordBatch> {
+    let mut left = 0;
+    let mut right = indices.len();
+
+    while left < right {
+        let mid = left + (right - left) / 2;
+        if indices[mid] < start {
+            left = mid + 1;
+        } else {
+            right = mid;
+        }
+    }
+    let start_index = left;
+
+    right = indices.len();
+
+    while left < right {
+        let mid = left + (right - left) / 2;
+        if indices[mid] <= end {
+            left = mid + 1;
+        } else {
+            right = mid;
+        }
+    }
+    let end_index = right;
+
+    if start_index < end_index {
+        batches[start_index..=end_index - 1].to_owned()
+    } else if start_index == end_index {
+        vec![batches[end_index - 1].to_owned()]
+    } else {
+        vec![]
+    }
+}
```

Comment:
Maybe we can use std RUST library here: `use std::collections::binary_search::partition_point;` ?
```suggestion

async fn fetch_partitions(
      batches: Vec<RecordBatch>,
      indices: &[u64],
      start: u64,
      end: u64,
  ) -> Vec<RecordBatch> {
      let start_idx = indices.partition_point(|&x| x < start);
      let end_idx = indices.partition_point(|&x| x <= end);

      if start_idx < end_idx && start_idx < batches.len() {
          batches[start_idx..end_idx.min(batches.len())].to_vec()
      } else {
          Vec::new()
      }
  }
```

Reply:
Nice. Refactored.